core/core.go

```diff
@@ -222,62 +258,235 @@ func NewTextile(ctx context.Context, conf Config) (*Textile, error) {
-	token, err := auth.AuthFromMD(ctx, "bearer")
-	if err != nil {
-		return nil, err
-	}
-	if token == t.threadsToken {
-		return ctx, nil
-	}
+func getService(method string) string {
+	return strings.Split(method, "/")[1]
+}
 
-	session, err := t.collections.Sessions.Get(ctx, token)
-	if err != nil {
-		return nil, status.Error(codes.Unauthenticated, "Invalid auth token")
+func (t *Textile) authFunc(ctx context.Context) (context.Context, error) {
+	method, _ := grpc.Method(ctx)
+	for _, ignored := range ignoreMethods {
+		if method == ignored {
+			return ctx, nil
+		}
 	}
-	if session.Expiry < int(time.Now().Unix()) {
-		return nil, status.Error(codes.Unauthenticated, "Expired auth token")
+
+	threadID, ok := common.ThreadIDFromMD(ctx)
+	if ok {
+		ctx = common.NewThreadIDContext(ctx, threadID)
 	}
-	user, err := t.collections.Users.Get(ctx, session.UserID)
-	if err != nil {
-		return nil, status.Error(codes.PermissionDenied, "User not found")
+	threadName, ok := common.ThreadNameFromMD(ctx)
+	if ok {
+		ctx = common.NewThreadNameContext(ctx, threadName)
 	}
-	proj, err := t.collections.Projects.Get(ctx, user.ProjectID)
+	threadToken, err := thread.NewTokenFromMD(ctx)
 	if err != nil {
-		return nil, status.Error(codes.PermissionDenied, "Project not found")
+		return nil, err
 	}
+	ctx = thread.NewTokenContext(ctx, threadToken)
 
-	if err := t.collections.Sessions.Touch(ctx, session); err != nil {
-		return nil, err
+	sid, ok := common.SessionFromMD(ctx)
+	if ok {
+		ctx = common.NewSessionContext(ctx, sid) // For additional requests
+		if sid == t.gatewaySession {
+			return ctx, nil
+		}
+		session, err := t.co.Sessions.Get(ctx, sid)
+		if err != nil {
+			return nil, status.Error(codes.Unauthenticated, "Invalid session")
+		}
+		if time.Now().After(session.ExpiresAt) {
+			return nil, status.Error(codes.Unauthenticated, "Expired session")
+		}
+		if err := t.co.Sessions.Touch(ctx, session.ID); err != nil {
+			return nil, err
+		}
+		ctx = c.NewSessionContext(ctx, session)
+
+		dev, err := t.co.Accounts.Get(ctx, session.Owner)
+		if err != nil {
+			return nil, status.Error(codes.NotFound, "User not found")
+		}
+		ctx = c.NewDevContext(ctx, dev)
+
+		orgSlug, ok := common.OrgSlugFromMD(ctx)
+		if ok {
+			isMember, err := t.co.Accounts.IsMember(ctx, orgSlug, dev.Key)
+			if err != nil {
+				return nil, err
+			}
+			if !isMember {
+				return nil, status.Error(codes.PermissionDenied, "User is not an org member")
+			} else {
+				org, err := t.co.Accounts.GetByUsername(ctx, orgSlug)
+				if err != nil {
+					return nil, status.Error(codes.NotFound, "Org not found")
+				}
+				ctx = c.NewOrgContext(ctx, org)
+				ctx = common.NewOrgSlugContext(ctx, orgSlug) // For additional requests
+				ctx = thread.NewTokenContext(ctx, org.Token)
+			}
+		} else {
+			ctx = thread.NewTokenContext(ctx, dev.Token)
+		}
+	} else if key, ok := common.APIKeyFromMD(ctx); ok {
+		ctx = common.NewAPIKeyContext(ctx, key)
```

Comment:
In this case, it's a user, they should have an API key and signature (sig is checked below).